File: shared/telemetry/metrics.py
"""
OpenTelemetry Metrics for SAHOOL Platform
==========================================

This module provides comprehensive metrics instrumentation for all
Python-based services in the SAHOOL agricultural platform.

Features:
- OpenTelemetry metrics (counters, histograms, gauges)
- Prometheus exporter for metrics scraping
- Request count, latency, error rate metrics
- Custom business metrics (fields, satellite requests, weather queries)
- Automatic metric labeling with service names

Usage:
    from shared.telemetry.metrics import init_metrics, track_request, track_business_metric

    # Initialize at application startup
    meter_provider = init_metrics(service_name="field_core")

    # Track HTTP requests
    track_request(method="GET", endpoint="/fields", status_code=200, duration=0.123)

    # Track business metrics
    track_business_metric("fields_created", value=1, labels={"user_id": "123"})

Author: SAHOOL Platform Team
Date: 2025-12-26
"""

import logging
import os
import time
from typing import Optional, Dict, Any
from contextlib import contextmanager

from opentelemetry import metrics
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.metrics.export import (
    PeriodicExportingMetricReader,
    ConsoleMetricExporter,
)
from opentelemetry.exporter.prometheus import PrometheusMetricReader
from opentelemetry.sdk.resources import Resource, SERVICE_NAME, SERVICE_VERSION, DEPLOYMENT_ENVIRONMENT
from prometheus_client import start_http_server, REGISTRY

logger = logging.getLogger(__name__)

# Global meter instance
_meter: Optional[metrics.Meter] = None
_meter_provider: Optional[MeterProvider] = None

# Metrics instruments
_request_counter: Optional[metrics.Counter] = None
_request_duration: Optional[metrics.Histogram] = None
_error_counter: Optional[metrics.Counter] = None
_business_counters: Dict[str, metrics.Counter] = {}
_business_histograms: Dict[str, metrics.Histogram] = {}
_business_gauges: Dict[str, metrics.ObservableGauge] = {}


def init_metrics(
    service_name: Optional[str] = None,
    service_version: Optional[str] = None,
    environment: Optional[str] = None,
    prometheus_port: int = 9090,
    enable_prometheus: bool = True,
    enable_console: bool = False,
) -> MeterProvider:
    """
    Initialize OpenTelemetry metrics with Prometheus exporter.

    Args:
        service_name: Name of the service (auto-detected from env if not provided)
        service_version: Version of the service
        environment: Deployment environment (development, staging, production)
        prometheus_port: Port for Prometheus metrics endpoint
        enable_prometheus: Enable Prometheus exporter
        enable_console: Enable console exporter for debugging

    Returns:
        MeterProvider instance
    """
    global _meter_provider, _meter
    global _request_counter, _request_duration, _error_counter

    # Auto-detect service name from environment
    if not service_name:
        service_name = os.getenv("OTEL_SERVICE_NAME") or os.getenv("SERVICE_NAME", "sahool-service")

    if not service_version:
        service_version = os.getenv("SERVICE_VERSION", "1.0.0")

    if not environment:
        environment = os.getenv("ENVIRONMENT", "development")

    # Create resource with service information
    resource = Resource.create({
        SERVICE_NAME: service_name,
        SERVICE_VERSION: service_version,
        DEPLOYMENT_ENVIRONMENT: environment,
        "service.namespace": "sahool",
        "service.instance.id": os.getenv("HOSTNAME", "unknown"),
    })

    # Configure metric readers
    metric_readers = []

    # Add Prometheus exporter
    if enable_prometheus or os.getenv("PROMETHEUS_METRICS_ENABLED", "true").lower() == "true":
        try:
            prometheus_reader = PrometheusMetricReader()
            metric_readers.append(prometheus_reader)

            # Start Prometheus HTTP server
            prometheus_port = int(os.getenv("PROMETHEUS_PORT", prometheus_port))
            start_http_server(port=prometheus_port, registry=REGISTRY)
            logger.info(f"Prometheus metrics endpoint started on port {prometheus_port}")
        except Exception as e:
            logger.error(f"Failed to configure Prometheus exporter: {e}")

    # Add console exporter for debugging
    if enable_console or os.getenv("OTEL_CONSOLE_EXPORT", "false").lower() == "true":
        console_reader = PeriodicExportingMetricReader(
            ConsoleMetricExporter(),
            export_interval_millis=60000,  # 1 minute
        )
        metric_readers.append(console_reader)
        logger.info("Console metrics exporter enabled")

    # Create meter provider
    _meter_provider = MeterProvider(
        resource=resource,
        metric_readers=metric_readers,
    )

    # Set global meter provider
    metrics.set_meter_provider(_meter_provider)

    # Get meter instance
    _meter = metrics.get_meter(
        name=service_name,
        version=service_version,
    )

    # Initialize standard HTTP metrics
    _request_counter = _meter.create_counter(
        name="http_requests_total",
        description="Total number of HTTP requests",
        unit="1",
    )

    _request_duration = _meter.create_histogram(
        name="http_request_duration_seconds",
        description="HTTP request duration in seconds",
        unit="s",
    )

    _error_counter = _meter.create_counter(
        name="http_errors_total",
        description="Total number of HTTP errors",
        unit="1",
    )

    logger.info(f"OpenTelemetry metrics initialized: service={service_name}, env={environment}")

    return _meter_provider


def get_meter(name: Optional[str] = None, version: Optional[str] = None) -> metrics.Meter:
    """
    Get a meter instance.

    Args:
        name: Meter name (defaults to service name)
        version: Meter version

    Returns:
        Meter instance
    """
    if not _meter_provider:
        logger.warning("Meter provider not initialized, using default")
        return metrics.get_meter(name or __name__, version=version)

    return metrics.get_meter(name or __name__, version=version)


def track_request(
    method: str,
    endpoint: str,
    status_code: int,
    duration: float,
    labels: Optional[Dict[str, str]] = None,
) -> None:
    """
    Track HTTP request metrics.

    Args:
        method: HTTP method (GET, POST, etc.)
        endpoint: API endpoint
        status_code: HTTP status code
        duration: Request duration in seconds
        labels: Additional labels
    """
    if not _request_counter or not _request_duration:
        return

    attributes = {
        "http.method": method,
        "http.endpoint": endpoint,
        "http.status_code": status_code,
    }

    if labels:
        attributes.update(labels)

    # Count request
    _request_counter.add(1, attributes=attributes)

    # Record duration
    _request_duration.record(duration, attributes=attributes)

    # Count errors (4xx, 5xx)
    if status_code >= 400 and _error_counter:
        error_type = "client_error" if status_code < 500 else "server_error"
        error_attributes = {**attributes, "error.type": error_type}
        _error_counter.add(1, attributes=error_attributes)


@contextmanager
def track_request_duration(method: str, endpoint: str, labels: Optional[Dict[str, str]] = None):
    """
    Context manager to track request duration.

    Args:
        method: HTTP method
        endpoint: API endpoint
        labels: Additional labels

    Example:
        with track_request_duration("GET", "/fields"):
            # ... process request
            pass
    """
    start_time = time.time()
    status_code = 200

    try:
        yield
    except Exception as e:
        status_code = 500
        raise
    finally:
        duration = time.time() - start_time
        track_request(method, endpoint, status_code, duration, labels)


def create_counter(
    name: str,
    description: str = "",
    unit: str = "1",
) -> metrics.Counter:
    """
    Create a custom counter metric.

    Args:
        name: Metric name
        description: Metric description
        unit: Metric unit

    Returns:
        Counter instance
    """
    meter = get_meter()
    return meter.create_counter(
        name=name,
        description=description,
        unit=unit,
    )


def create_histogram(
    name: str,
    description: str = "",
    unit: str = "1",
) -> metrics.Histogram:
    """
    Create a custom histogram metric.

    Args:
        name: Metric name
        description: Metric description
        unit: Metric unit

    Returns:
        Histogram instance
    """
    meter = get_meter()
    return meter.create_histogram(
        name=name,
        description=description,
        unit=unit,
    )


def create_up_down_counter(
    name: str,
    description: str = "",
    unit: str = "1",
) -> metrics.UpDownCounter:
    """
    Create a custom up-down counter metric (can go up and down).

    Args:
        name: Metric name
        description: Metric description
        unit: Metric unit

    Returns:
        UpDownCounter instance
    """
    meter = get_meter()
    return meter.create_up_down_counter(
        name=name,
        description=description,
        unit=unit,
    )


def track_business_metric(
    metric_name: str,
    value: float = 1,
    metric_type: str = "counter",
    labels: Optional[Dict[str, Any]] = None,
    description: str = "",
    unit: str = "1",
) -> None:
    """
    Track custom business metrics.

    Args:
        metric_name: Name of the metric
        value: Metric value
        metric_type: Type of metric (counter, histogram)
        labels: Metric labels
        description: Metric description
        unit: Metric unit

    Example:
        # Track fields created
        track_business_metric("fields_created", value=1, labels={"user_id": "123"})

        # Track satellite request latency
        track_business_metric(
            "satellite_request_duration",
            value=2.5,
            metric_type="histogram",
            labels={"provider": "sentinel"}
        )
    """
    global _business_counters, _business_histograms

    attributes = labels or {}

    if metric_type == "counter":
        if metric_name not in _business_counters:
            _business_counters[metric_name] = create_counter(
                name=metric_name,
                description=description or f"Business metric: {metric_name}",
                unit=unit,
            )
        _business_counters[metric_name].add(value, attributes=attributes)

    elif metric_type == "histogram":
        if metric_name not in _business_histograms:
            _business_histograms[metric_name] = create_histogram(
                name=metric_name,
                description=description or f"Business metric: {metric_name}",
                unit=unit,
            )
        _business_histograms[metric_name].record(value, attributes=attributes)


# Predefined business metrics for SAHOOL services
class SahoolMetrics:
    """Standard business metrics for SAHOOL platform."""

    @staticmethod
    def track_field_created(user_id: str, field_type: str = "agricultural"):
        """Track field creation."""
        track_business_metric(
            "sahool_fields_created_total",
            value=1,
            labels={"user_id": user_id, "field_type": field_type},
            description="Total number of fields created",
        )

    @staticmethod
    def track_satellite_request(provider: str, status: str, duration: float):
        """Track satellite imagery request."""
        track_business_metric(
            "sahool_satellite_requests_total",
            value=1,
            labels={"provider": provider, "status": status},
            description="Total number of satellite requests",
        )
        track_business_metric(
            "sahool_satellite_request_duration_seconds",
            value=duration,
            metric_type="histogram",
            labels={"provider": provider},
            description="Satellite request duration in seconds",
            unit="s",
        )

    @staticmethod
    def track_weather_query(provider: str, location: str, status: str):
        """Track weather data query."""
        track_business_metric(
            "sahool_weather_queries_total",
            value=1,
            labels={"provider": provider, "location": location, "status": status},
            description="Total number of weather queries",
        )

    @staticmethod
    def track_ndvi_calculation(field_id: str, duration: float):
        """Track NDVI calculation."""
        track_business_metric(
            "sahool_ndvi_calculations_total",
            value=1,
            labels={"field_id": field_id},
            description="Total number of NDVI calculations",
        )
        track_business_metric(
            "sahool_ndvi_calculation_duration_seconds",
            value=duration,
            metric_type="histogram",
            labels={"field_id": field_id},
            description="NDVI calculation duration in seconds",
            unit="s",
        )

    @staticmethod
    def track_ai_recommendation(advisor_type: str, crop_type: str, duration: float):
        """Track AI advisory recommendation."""
        track_business_metric(
            "sahool_ai_recommendations_total",
            value=1,
            labels={"advisor_type": advisor_type, "crop_type": crop_type},
            description="Total number of AI recommendations",
        )
        track_business_metric(
            "sahool_ai_recommendation_duration_seconds",
            value=duration,
            metric_type="histogram",
            labels={"advisor_type": advisor_type},
            description="AI recommendation duration in seconds",
            unit="s",
        )

    @staticmethod
    def track_iot_reading(sensor_type: str, field_id: str, value: float):
        """Track IoT sensor reading."""
        track_business_metric(
            "sahool_iot_readings_total",
            value=1,
            labels={"sensor_type": sensor_type, "field_id": field_id},
            description="Total number of IoT sensor readings",
        )

    @staticmethod
    def track_notification_sent(notification_type: str, channel: str, status: str):
        """Track notification delivery."""
        track_business_metric(
            "sahool_notifications_sent_total",
            value=1,
            labels={
                "notification_type": notification_type,
                "channel": channel,
                "status": status,
            },
            description="Total number of notifications sent",
        )

    @staticmethod
    def track_irrigation_event(field_id: str, event_type: str, water_amount: float):
        """Track irrigation event."""
        track_business_metric(
            "sahool_irrigation_events_total",
            value=1,
            labels={"field_id": field_id, "event_type": event_type},
            description="Total number of irrigation events",
        )
        track_business_metric(
            "sahool_irrigation_water_liters",
            value=water_amount,
            metric_type="histogram",
            labels={"field_id": field_id},
            description="Irrigation water amount in liters",
            unit="L",
        )

    @staticmethod
    def track_crop_health_analysis(crop_type: str, health_score: float, duration: float):
        """Track crop health analysis."""
        track_business_metric(
            "sahool_crop_health_analyses_total",
            value=1,
            labels={"crop_type": crop_type},
            description="Total number of crop health analyses",
        )
        track_business_metric(
            "sahool_crop_health_score",
            value=health_score,
            metric_type="histogram",
            labels={"crop_type": crop_type},
            description="Crop health score (0-100)",
            unit="score",
        )

    @staticmethod
    def track_yield_prediction(crop_type: str, predicted_yield: float, duration: float):
        """Track yield prediction."""
        track_business_metric(
            "sahool_yield_predictions_total",
            value=1,
            labels={"crop_type": crop_type},
            description="Total number of yield predictions",
        )
        track_business_metric(
            "sahool_predicted_yield_kg",
            value=predicted_yield,
            metric_type="histogram",
            labels={"crop_type": crop_type},
            description="Predicted yield in kilograms",
            unit="kg",
        )

    @staticmethod
    def track_marketplace_transaction(transaction_type: str, amount: float, status: str):
        """Track marketplace transaction."""
        track_business_metric(
            "sahool_marketplace_transactions_total",
            value=1,
            labels={"transaction_type": transaction_type, "status": status},
            description="Total number of marketplace transactions",
        )
        track_business_metric(
            "sahool_marketplace_revenue",
            value=amount,
            metric_type="histogram",
            labels={"transaction_type": transaction_type},
            description="Marketplace transaction revenue",
            unit="currency",
        )


__all__ = [
    "init_metrics",
    "get_meter",
    "track_request",
    "track_request_duration",
    "create_counter",
    "create_histogram",
    "create_up_down_counter",
    "track_business_metric",
    "SahoolMetrics",
]

File: shared/telemetry/metrics.ts
/**
 * Prometheus Metrics for SAHOOL Platform (TypeScript/NestJS)
 * ===========================================================
 *
 * This module provides comprehensive metrics instrumentation for all
 * TypeScript-based services (NestJS) in the SAHOOL agricultural platform.
 *
 * Features:
 * - Prometheus client integration
 * - Request count, latency, error rate metrics
 * - Custom business metrics (fields, notifications, chat messages)
 * - Automatic metric labeling
 * - Metrics endpoint (/metrics)
 *
 * Usage:
 *     import { initMetrics, trackRequest, trackBusinessMetric } from './shared/telemetry/metrics';
 *
 *     // Initialize at application startup
 *     initMetrics({ serviceName: 'chat-service' });
 *
 *     // Track HTTP requests
 *     trackRequest('GET', '/messages', 200, 0.123);
 *
 *     // Track business metrics
 *     trackBusinessMetric('messages_sent', 1, { channel: 'field-chat' });
 *
 * Author: SAHOOL Platform Team
 * Date: 2025-12-26
 */

import * as promClient from "prom-client";
import { INestApplication } from "@nestjs/common";

let register: promClient.Registry;
let metricsInitialized = false;

// Standard HTTP metrics
let httpRequestsTotal: promClient.Counter<string>;
let httpRequestDuration: promClient.Histogram<string>;
let httpErrorsTotal: promClient.Counter<string>;

// Business metrics storage
const businessCounters: Map<string, promClient.Counter<string>> = new Map();
const businessHistograms: Map<string, promClient.Histogram<string>> = new Map();
const businessGauges: Map<string, promClient.Gauge<string>> = new Map();

export interface MetricsConfig {
  serviceName?: string;
  serviceVersion?: string;
  environment?: string;
  defaultLabels?: Record<string, string>;
  collectDefaultMetrics?: boolean;
}

/**
 * Initialize Prometheus metrics
 */
export function initMetrics(config: MetricsConfig = {}): promClient.Registry {
  if (metricsInitialized) {
    console.warn("Metrics already initialized");
    return register;
  }

  // Auto-detect service name from environment
  const serviceName =
    config.serviceName ||
    process.env.OTEL_SERVICE_NAME ||
    process.env.SERVICE_NAME ||
    "sahool-service";

  const serviceVersion =
    config.serviceVersion || process.env.SERVICE_VERSION || "1.0.0";
  const environment =
    config.environment || process.env.ENVIRONMENT || "development";

  // Create new registry
  register = new promClient.Registry();

  // Set default labels
  const defaultLabels = {
    service: serviceName,
    version: serviceVersion,
    environment,
    ...config.defaultLabels,
  };
  register.setDefaultLabels(defaultLabels);

  // Collect default Node.js metrics
  if (config.collectDefaultMetrics ?? true) {
    promClient.collectDefaultMetrics({ register });
  }

  // Initialize standard HTTP metrics
  httpRequestsTotal = new promClient.Counter({
    name: "http_requests_total",
    help: "Total number of HTTP requests",
    labelNames: ["method", "endpoint", "status_code"],
    registers: [register],
  });

  httpRequestDuration = new promClient.Histogram({
    name: "http_request_duration_seconds",
    help: "HTTP request duration in seconds",
    labelNames: ["method", "endpoint", "status_code"],
    buckets: [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10],
    registers: [register],
  });

  httpErrorsTotal = new promClient.Counter({
    name: "http_errors_total",
    help: "Total number of HTTP errors",
    labelNames: ["method", "endpoint", "status_code", "error_type"],
    registers: [register],
  });

  metricsInitialized = true;

  console.log(
    `Prometheus metrics initialized: service=${serviceName}, env=${environment}`,
  );

  return register;
}

/**
 * Get the metrics registry
 */
export function getRegistry(): promClient.Registry {
  if (!register) {
    throw new Error("Metrics not initialized. Call initMetrics() first.");
  }
  return register;
}

/**
 * Get metrics endpoint handler
 */
export async function getMetrics(): Promise<string> {
  return register.metrics();
}

/**
 * Track HTTP request metrics
 */
export function trackRequest(
  method: string,
  endpoint: string,
  statusCode: number,
  duration: number,
  labels?: Record<string, string>,
): void {
  if (!metricsInitialized) return;

  const metricLabels = {
    method,
    endpoint,
    status_code: statusCode.toString(),
    ...labels,
  };

  // Count request
  httpRequestsTotal.inc(metricLabels);

  // Record duration
  httpRequestDuration.observe(metricLabels, duration);

  // Count errors (4xx, 5xx)
  if (statusCode >= 400) {
    const errorType = statusCode < 500 ? "client_error" : "server_error";
    httpErrorsTotal.inc({
      ...metricLabels,
      error_type: errorType,
    });
  }
}

/**
 * Create a custom counter metric
 */
export function createCounter(
  name: string,
  help: string,
  labelNames: string[] = [],
): promClient.Counter<string> {
  if (!register) {
    throw new Error("Metrics not initialized");
  }

  const counter = new promClient.Counter({
    name,
    help,
    labelNames,
    registers: [register],
  });

  return counter;
}

/**
 * Create a custom histogram metric
 */
export function createHistogram(
  name: string,
  help: string,
  labelNames: string[] = [],
  buckets?: number[],
): promClient.Histogram<string> {
  if (!register) {
    throw new Error("Metrics not initialized");
  }

  const histogram = new promClient.Histogram({
    name,
    help,
    labelNames,
    buckets: buckets || [0.001, 0.005, 0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10],
    registers: [register],
  });

  return histogram;
}

/**
 * Create a custom gauge metric
 */
export function createGauge(
  name: string,
  help: string,
  labelNames: string[] = [],
): promClient.Gauge<string> {
  if (!register) {
    throw new Error("Metrics not initialized");
  }

  const gauge = new promClient.Gauge({
    name,
    help,
    labelNames,
    registers: [register],
  });

  return gauge;
}

/**
 * Track custom business metrics
 */
export function trackBusinessMetric(
  metricName: string,
  value: number = 1,
  metricType: "counter" | "histogram" | "gauge" = "counter",
  labels?: Record<string, string>,
  help?: string,
): void {
  if (!metricsInitialized) return;

  const labelNames = labels ? Object.keys(labels) : [];

  if (metricType === "counter") {
    let counter = businessCounters.get(metricName);
    if (!counter) {
      counter = createCounter(
        metricName,
        help || `Business metric: ${metricName}`,
        labelNames,
      );
      businessCounters.set(metricName, counter);
    }
    counter.inc(labels || {}, value);
  } else if (metricType === "histogram") {
    let histogram = businessHistograms.get(metricName);
    if (!histogram) {
      histogram = createHistogram(
        metricName,
        help || `Business metric: ${metricName}`,
        labelNames,
      );
      businessHistograms.set(metricName, histogram);
    }
    histogram.observe(labels || {}, value);
  } else if (metricType === "gauge") {
    let gauge = businessGauges.get(metricName);
    if (!gauge) {
      gauge = createGauge(
        metricName,
        help || `Business metric: ${metricName}`,
        labelNames,
      );
      businessGauges.set(metricName, gauge);
    }
    gauge.set(labels || {}, value);
  }
}

/**
 * Middleware to track request duration
 */
export function trackRequestDuration(
  method: string,
  endpoint: string,
  labels?: Record<string, string>,
) {
  const startTime = Date.now();

  return (statusCode: number = 200) => {
    const duration = (Date.now() - startTime) / 1000;
    trackRequest(method, endpoint, statusCode, duration, labels);
  };
}

/**
 * Standard business metrics for SAHOOL platform
 */
export class SahoolMetrics {
  /**
   * Track chat message sent
   */
  static trackMessageSent(
    channel: string,
    messageType: string,
    userId: string,
  ): void {
    trackBusinessMetric(
      "sahool_messages_sent_total",
      1,
      "counter",
      { channel, message_type: messageType, user_id: userId },
      "Total number of chat messages sent",
    );
  }

  /**
   * Track notification sent
   */
  static trackNotificationSent(
    notificationType: string,
    channel: string,
    status: string,
  ): void {
    trackBusinessMetric(
      "sahool_notifications_sent_total",
      1,
      "counter",
      { notification_type: notificationType, channel, status },
      "Total number of notifications sent",
    );
  }

  /**
   * Track notification delivery latency
   */
  static trackNotificationLatency(
    notificationType: string,
    duration: number,
  ): void {
    trackBusinessMetric(
      "sahool_notification_duration_seconds",
      duration,
      "histogram",
      { notification_type: notificationType },
      "Notification delivery duration in seconds",
    );
  }

  /**
   * Track WebSocket connections
   */
  static trackWebSocketConnection(event: "connect" | "disconnect"): void {
    const gauge = businessGauges.get("sahool_websocket_connections");
    if (gauge) {
      if (event === "connect") {
        gauge.inc();
      } else {
        gauge.dec();
      }
    } else {
      const newGauge = createGauge(
        "sahool_websocket_connections",
        "Current number of WebSocket connections",
      );
      businessGauges.set("sahool_websocket_connections", newGauge);
      newGauge.set(event === "connect" ? 1 : 0);
    }
  }

  /**
   * Track billing transaction
   */
  static trackBillingTransaction(
    transactionType: string,
    amount: number,
    status: string,
  ): void {
    trackBusinessMetric(
      "sahool_billing_transactions_total",
      1,
      "counter",
      { transaction_type: transactionType, status },
      "Total number of billing transactions",
    );

    trackBusinessMetric(
      "sahool_billing_revenue",
      amount,
      "histogram",
      { transaction_type: transactionType },
      "Billing transaction revenue",
    );
  }

  /**
   * Track marketplace listing
   */
  static trackMarketplaceListing(
    listingType: string,
    action: "create" | "update" | "delete",
  ): void {
    trackBusinessMetric(
      "sahool_marketplace_listings_total",
      1,
      "counter",
      { listing_type: listingType, action },
      "Total number of marketplace listings",
    );
  }

  /**
   * Track inventory operation
   */
  static trackInventoryOperation(
    operationType: string,
    itemType: string,
    quantity: number,
  ): void {
    trackBusinessMetric(
      "sahool_inventory_operations_total",
      1,
      "counter",
      { operation_type: operationType, item_type: itemType },
      "Total number of inventory operations",
    );

    trackBusinessMetric(
      "sahool_inventory_quantity",
      quantity,
      "gauge",
      { item_type: itemType },
      "Current inventory quantity",
    );
  }

  /**
   * Track equipment usage
   */
  static trackEquipmentUsage(equipmentType: string, usageHours: number): void {
    trackBusinessMetric(
      "sahool_equipment_usage_hours",
      usageHours,
      "histogram",
      { equipment_type: equipmentType },
      "Equipment usage in hours",
    );
  }

  /**
   * Track task completion
   */
  static trackTaskCompletion(
    taskType: string,
    status: string,
    duration: number,
  ): void {
    trackBusinessMetric(
      "sahool_tasks_completed_total",
      1,
      "counter",
      { task_type: taskType, status },
      "Total number of tasks completed",
    );

    trackBusinessMetric(
      "sahool_task_duration_seconds",
      duration,
      "histogram",
      { task_type: taskType },
      "Task completion duration in seconds",
    );
  }

  /**
   * Track chat session duration
   */
  static trackChatSession(duration: number, participants: number): void {
    trackBusinessMetric(
      "sahool_chat_session_duration_seconds",
      duration,
      "histogram",
      {},
      "Chat session duration in seconds",
    );

    trackBusinessMetric(
      "sahool_chat_participants",
      participants,
      "histogram",
      {},
      "Number of chat participants",
    );
  }
}

/**
 * NestJS Interceptor for automatic request tracking
 */
export class MetricsInterceptor {
  async intercept(context: any, next: any): Promise<any> {
    const request = context.switchToHttp().getRequest();
    const response = context.switchToHttp().getResponse();

    const startTime = Date.now();
    const method = request.method;
    const endpoint = request.route?.path || request.url;

    try {
      const result = await next.handle().toPromise();
      const duration = (Date.now() - startTime) / 1000;
      trackRequest(method, endpoint, response.statusCode || 200, duration);
      return result;
    } catch (error) {
      const duration = (Date.now() - startTime) / 1000;
      trackRequest(method, endpoint, 500, duration);
      throw error;
    }
  }
}

export default {
  initMetrics,
  getRegistry,
  getMetrics,
  trackRequest,
  trackRequestDuration,
  createCounter,
  createHistogram,
  createGauge,
  trackBusinessMetric,
  SahoolMetrics,
  MetricsInterceptor,
};
